__uk_init_tls.c

```diff
@@ -0,0 +1,195 @@
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <uk/tcb_impl.h>
+#include <elf.h>
+#include <limits.h>
+#include <sys/mman.h>
+#include <string.h>
+#include <stddef.h>
+#include "pthread_impl.h"
+#include "libc.h"
+#include "atomic.h"
+#include "syscall.h"
+
+#include <uk/alloc.h>
+#include <uk/assert.h>
+
+/*
+ * This glue code source is meant to replace `__libc_start_main()`
+ * and tls related stuff.
+ * If the pthread API is used, musl will allocate the TLS and will
+ * provide a pointer to it in the `clone()` syscall. Musl will allocate
+ * in fact a contiguous region which will have the following layout (Please
+ * refer to the `pthread_create.c` file):
+ * map -----------------------------------------------------------------
+ *                     ^
+ *                     |
+ *                     | 4096      GUARD PAGE
+ *                     |
+ *                     v
+ * map + guard ---------------------------------------------------------
+ *                     ^
+ *                     |
+ *                     |
+ *                     |           STACK
+ *                     |
+ *                     |
+ *                     |
+ *                     v
+ * stack ---------------------------------------------------------------
+ *                     ^           ^
+ *                     |           |       TLS SPACE
+ *                     |           v
+ *               new ->|------------------------------------------------
+ *                     |           ^
+ *                     |           |
+ *                     |           |
+ *                     |           | 280   pthread structure
+ *                     |           |
+ *                     v           v
+ * tsd -----------------------------------------------------------------
+ *                     ^
+ *                     | 1024
+ *                     v
+ * map + size ----------------------------------------------------------
+ */
+
+
+void *__uk_copy_tls(unsigned char *mem)
+{
+	pthread_t td;
+	void *tls_area;
+
+	mem -= (uintptr_t)mem & (libc.tls_align-1);
+	tls_area = mem;
+	ukarch_tls_area_init(tls_area);
+
+	td = (pthread_t) ukarch_tls_tlsp(tls_area);
+	td->dtv = td->dtv_copy = tls_area;
+
+	return td;
+}
+
+static int __uk_init_tp(void *p)
+{
+	pthread_t td = p;
+
+	td->self = td;
+	/*
+	 * Set the `$fs` register for the current thread.
+	 * In the original code of musl this will use an `arch_prtcl`
+	 * syscall to fill the `$fs` register.
+	 */
+	ukplat_tlsp_set(TP_ADJ(p));
+	libc.can_do_threads = 1;
+	/*
+	 * The original musl code will invoke here a `SYS_set_tid_address`
+	 * syscall, to set the tid user space address in the Kernel.
+	 * FIXME: Currently this does not return the tid assigned for the caller,
+	 * it returns an error code (-95) because probably there is no tid assigned
+	 * at this stage. It is not a really big problem right now.
+	 */
+	td->tid = uk_syscall_r_set_tid_address(&td->tid);
```

Comment:
This effectively sets the tid of the main thread to -95. This tid is then used, in the implementation of recursive mutexes, causing `pthread_mutex_unlock` to fail.

```suggestion
	 * it returns an error code (-95, -ENOTSUP) because posix_process_init has not been
	 * called at this stage, but will be called via uk_late_initcall.
	 * It is not a really big problem right now. Since this is the main thread,
	 * nobody should ever wait for it, and we can just assume thread id 0.
	 */
	td->tid = uk_syscall_r_set_tid_address(&td->tid);
	if (td->tid < 0) {
		td->tid = 0;
	}
```

Reply:
Hello, Robert. Thank you for your comment. You are right, the tid can be set to 0 here.